connectivity/curl/src/multi_performer.cpp

```diff
@@ -0,0 +1,131 @@
+// Copyright Antioch. All rights reserved.
+
+#include "multi_performer.h"
+
+#include <cstring>
+#include <iostream>
+#include <thread>
+
+#include "transfer.h"
+
+namespace antioch::connectivity::curl_transfer {
+
+MultiPerformer::MultiPerformer(CURLM* multi_handle) : multi_handle(multi_handle) {
+  std::thread thr(&MultiPerformer::run, this);
+  thr.detach();
+}
+
+MultiPerformer::~MultiPerformer() {
+  std::lock_guard<std::mutex> l(mtx);
+  std::cout << "got lock" << std::endl;
+  shutdown = true;
+  flag = true;
+  cv.notify_all();
+  for (const auto& pv : handle_objs) {
+    curl_multi_remove_handle(multi_handle, pv.first);
+  }
+  std::cout << "done dtor" << std::endl;
+}
+
+static size_t write_cb(void *data, size_t size, size_t nmemb, void *clientp) {
+  std::cout << "Called write_cb" << std::endl;
+  size_t realsize = size * nmemb;
+  auto mem = (EasyObj*)clientp;
+  auto ptr = (char*)realloc(mem->buf, mem->size + realsize + 1);
+  if(ptr == nullptr) {
+    return 0;
+  }
+ 
+  mem->buf = ptr;
+  std::memcpy(&(mem->buf[mem->size]), data, realsize);
+  mem->size += realsize;
+  mem->buf[mem->size] = 0; // null-terminate string
+ 
+  return realsize;
+}
+
+void MultiPerformer::queue_transfer(CURL* easy_handle, void (*finish_cb)(std::string)) {
+  std::scoped_lock<std::mutex> l(mtx);
+  handle_objs[easy_handle] = {finish_cb, 0, nullptr};
+  curl_easy_setopt(easy_handle, CURLOPT_WRITEFUNCTION, write_cb);
+  curl_easy_setopt(easy_handle, CURLOPT_WRITEDATA, &handle_objs[easy_handle]);
+  auto res = curl_multi_add_handle(multi_handle, easy_handle);
+  if (res != CURLM_OK) {
+    std::string msg = "curl_multi_add_handle returned " + std::to_string(res);
+    handle_objs.erase(handle_objs.find(easy_handle));
+    throw LibCurlInternalException(msg);
+  }
+  if (!flag) {
+    flag = true;
+    cv.notify_all();
+  }
+}
+
+void MultiPerformer::run() {
+  while (1) {
```

Comment:
Doesn't this need an exit condition

Reply:
exit condition is `if (shutdown)`

Reply:
But that returns not breaks, the code below this is not reachable right?